### ISSUE TITLE
Ensure Azure Blob Storage properly cleans up cancelled multipart uploads

### ISSUE THREAD

**Issue by danmason** [xtdb/crux#3629]
Have seen a few IOOBEs with multi node auctionmark runs, and examining the files causing issues, saw that they were `0 bytes`. Also observed that we were interrupting mid compaction on the "load phase" node, which was causing multiparts to be interrupted/aborted.

Adding in some extra logging, observed the following:
```
14:00:29.511 [xtdb.compactor-22] INFO  xtdb.buffer-pool - Error caught in upload-multipart-buffers - aborting multipart upload of tables/item/data/log-l01-fr4c870a-nr5113962-rs4b000.arrow
14:00:29.544 [xtdb.compactor-22] ERROR xtdb.compactor - Error running compaction job.
reactor.core.Exceptions$ReactiveException: java.lang.InterruptedException
    at reactor.core.Exceptions.propagate(Exceptions.java:396)
    at reactor.core.publisher.BlockingSingleSubscriber.blockingGet(BlockingSingleSubscriber.java:92)
    at reactor.core.publisher.Mono.block(Mono.java:1742)
    at com.azure.storage.common.implementation.StorageImplUtils.blockWithOptionalTimeout(StorageImplUtils.java:146)
    at com.azure.storage.blob.specialized.BlockBlobClient.stageBlockWithResponse(BlockBlobClient.java:787)
    at com.azure.storage.blob.specialized.BlockBlobClient.stageBlock(BlockBlobClient.java:704)
    at xtdb.azure.object_store.MultipartUpload.uploadPart(object_store.clj:77)
    at xtdb.buffer_pool$upload_multipart_buffers.invokeStatic(buffer_pool.clj:254)
    at xtdb.buffer_pool$upload_multipart_buffers.invoke(buffer_pool.clj:244)
    at xtdb.buffer_pool$upload_arrow_file$fn__8443.invoke(buffer_pool.clj:304)
    at xtdb.buffer_pool$upload_arrow_file.invokeStatic(buffer_pool.clj:303)
    at xtdb.buffer_pool$upload_arrow_file.invoke(buffer_pool.clj:297)
    at xtdb.buffer_pool.RemoteBufferPool$reify__8500.end(buffer_pool.clj:372)
    at xtdb.trie$open_trie_writer$reify__15748.end(trie.clj:178)
    at xtdb.Compactor.writeRelation(Compactor.kt:193)
    at xtdb.compactor$exec_compaction_job_BANG_$fn__16075$fn__16076$fn__16085$fn__16086$fn__16087.invoke(compactor.clj:142)
    at xtdb.compactor$exec_compaction_job_BANG_$fn__16075$fn__16076$fn__16085$fn__16086.invoke(compactor.clj:139)
    at xtdb.compactor$exec_compaction_job_BANG_$fn__16075$fn__16076$fn__16085.invoke(compactor.clj:135)
    at xtdb.compactor$exec_compaction_job_BANG_$fn__16075$fn__16076.invoke(compactor.clj:135)
    at xtdb.compactor$exec_compaction_job_BANG_$fn__16075.invoke(compactor.clj:124)
    at xtdb.compactor$exec_compaction_job_BANG_.invokeStatic(compactor.clj:124)
    at xtdb.compactor$exec_compaction_job_BANG_.invoke(compactor.clj:119)
    at xtdb.compactor$eval16191$fn__16193$compactor_task__16222$fn__16224.invoke(compactor.clj:273)
    at xtdb.compactor$eval16191$fn__16193$compactor_task__16222.invoke(compactor.clj:271)
    at xtdb.compactor$eval16191$fn__16193$fn__16227$fn__16228$fn__16229$fn__16235.invoke(compactor.clj:308)
    at clojure.lang.AFn.run(AFn.java:22)
    at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:572)
    at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:317)
    at java.base/java.lang.VirtualThread.run(VirtualThread.java:329)
```

Essentially:
- Thread gets interrupted on process close.
- The `upload-multipart-buffers` call catches the error (hence the 'error caught' log), and calls to `abort` on the multipart upload.
- No particular indication of errors/issues in that abort call.
- Looking at blob storage, can STILL see a file for `log-l01-fr4c870a-nr5113962-rs4b000.arrow` - of size 0 bytes.
- This invalid 0 byte arrow file causes later IOOBEs when interracting with the `items` table.

Should add some stronger testing around `abort` calls and ensure that no files are ever left hanging around in the case of calling an `abort` operation. Could well be some inherent async behaviour within there which we're breaking on from a lack of `await` when closing the process - need to investigate further.

TODO:
- [x] Better handle ReactiveException - rethrow a standard InterruptedException and ensure that the normal IE handling is triggered. Ensure that the process (where possible) isn't terminated during a critical section.
- [x] Document up contracted and observed Azure behaviour for multipart uploads, particularly around the atomicity of the various operations, and what happens if multiple clients are uploading the same file concurrently.
- [ ] Consider uploading to a tmp file in the remote object store. Iff that upload is successful, atomically move the tmp file into the right location.

**Comment by danmason** [comment]
Prior to giving context around Azure operations, I wanted to spend some time writing up our process for handling multipart uploads within Azure, and what triggers them to happen.

Currently, within the compaction job, we trigger a call which eventually calls down to the arrow writer within buffer pool. I'll focus on the behaviour from there. 

Essentially, we start by a call to `upload-arrow-file` - which in itself first decides whether or not to use `putObject` or start a multipart upload. If the object store supports multipart (as we do for azure), and size of the file is higher than a minimum threshold (5MB), we call to `upload-multipart-buffers`, splitting our arrow buffer into a set of parts.

`upload-multipart-buffers` makes use of our multipart interface to go through the process of a uploading a file in parts:
- We start by calling to `.startMultipart` for the current filename/key on the object store.
  - We call a function to create the multipart upload - starting by creating a **BlockBlobClient** for the current blob/file.  
  - We create our **MultipartUpload** record which we will call our multipart methods on. We pass this the blockblobclient and an empty arraylist.
- With the **MultipartUpload** record in hand, we start looping through the parts we have made from our original buffer for the arrow file. While we still have parts, we take the first part off of the head of the queue and call to `.uploadPart` on the **MultipartUpload**.
  - We create a unique **BlockId** to identify the current block.
  - We convert the part into an azure **BinaryData** object.
  - We call to "stage" the current block on the **BlockBlobClient** - passing in the **BlockId** and the **BinaryData* - essentially, we're creating a new block as part of a block blob's "staging area" (to be eventually committed as part of the full block blob).
  - We add the **BlockId** to our array list (in sequential order) - such that we can commit all of these blocks in the correct order on completion.
- As we return a completeable future from `uploadPart` (though to note: we do seem to use it fairly synchronously, returning completed futures _after_ we've done the work within them), we submit all of the parts of the arrow buffer and then await all of the futures to be complete.
- Once the loop over the parts is complete, we call to `complete` the **MultipartUpload**:
  - We call `commitBlockList` on the  **BlockBlobClient** - passing in our list of **staged BlockIds** - this writes the blob to the object store container by specifying the list of block IDs that make up the blob (In order to be written as part of a blob, a block must have been successfully written to the server prior).
  - If there is already a blob in place - we do no further work (ie, cancelling the multipart upload).
  - If this catches an `InterruptedException` (inside of Reactive - we have to `unwrap` the exception) - we rethrow the interrupted exception, which falls through to the error handling within `upload-multipart-buffers`

Within `upload-multipart-buffers`, if we catch an exception either within the loop or in `complete` on the **MultipartUpload**, we do the following:
- We log out that an error was caught and we are aborting the multipart upload.
- We call to `.abort` the **MultipartUpload** - this does the following, currently (as was the recommended for "clearing up" staged blocks:
  - We call to `commitBlockList` on the **BlockBlobClient** - this writes an empty (0 byte/0 blocks) blob to the object store.
  - Following this, we call `deleteIfExists` on the  **BlockBlobClient** - deleting the empty file. Essentially, clearly up both the staging area and the created file.
  - We catch an exception in this place - checking if we attempted to `commitBlockList` on a blob that already exists. In this case, as with complete, we do no further work (ie, cancel the multipart upload & abort process).
  - We currently rethrow any other unidentified `BlobStorageException`, though we're not catching any other throwables in there.
- If any errors are thrown during the `abort` process, we append it to the exception we received from the loop and return it.

**Comment by danmason** [comment]
Now, following on from the above, it's worth mentioning that we believe we have identified the actual problematic area. Essentially, the order of things as they happen:
- At the end of the benchmark, we exit the process - throwing an interrupt signal to our threads.
- This interrupts us in the middle of uploading one of the multipart arrow files for a table.
- Essentially, we go down the "error" path within `upload-multipart-buffers`, going to `abort` the multipart upload.
  - This is supposed to commit an empty blocklist and then delete the file, but in the error case we are left with an empty file (which causes IOOBEs on restart).

Investigating this with some extra logging, we essentially realized that:
- The call to `commitBlockList` was failing - ALSO returning a ReactiveException wrapped InterruptedException.
- As such, we wouldn't reach the `deleteIfExists` call - so essentially the `commitBlockList` WAS writing an empty file but also throwing an exception, causing our problems.

I'd say this comes under point "Better handle ReactiveException" of our TODO list - ie, one last unhandled ReactiveException is causing us problems. Should we handle this properly - we should be able to delete the unwanted file. That being said, the fact we still have to worry about "blob already exists" in the abort case does give a slight smell of "we should write to another file & copy" (though the lack of atomic move/copy makes this more difficult)

**Comment by danmason** [comment]
So, from our process of uploading multipart files, we use the following Azure Operations:

- `stageBlock`
  - in the case of multiple nodes writing to the same file, they'd all be staging different blocks/blockids anyway. When a block is successfully uploaded, it is stored temporarily in Azure but is not yet part of the final blob until the commitBlockList operation is executed (and if it is passed in to that commitBlockList).
- `commitBlockList`
  - We've seen in practise that this isn't very atomic on the java side - ie, we've had an interrupted exception thrown partway through and it STILL uploaded a 0 byte file. In theory if multiple nodes called to `commitBlockList` at the same time, the first one would be the one we _try_ to apply and the other two would get back "blob already exists" - even if that first one doesn't succeed.
- `deleteIfExists` (in the case of an abort) - by my understanding, we shouldn't run into a situation where more than one node can run this in their abort (since if two attempted the operation simultaneously only the first would get to the point of `deleteIfExists` since the other would fail with a "blob already exists"). Returns a true/false if the file was deleted.

**Comment by danmason** [comment]
Spending a bit of time thinking about this, given the lack of any form of atomic move and lack of atomic behaviour from the methods, I'm not particularly convinced of any particular approach at the moment - ie, if the copy isnt atomic either (and seems to have its OWN limitations around file sizes and the like), I feel like we risk running into very similar issues doing a "temp file that gets copied" approach. 

Going to spend some time reading around what exactly Azure expects from people in these scenarios, so I'm going to dive a bit deeper into their docs - see ["Managing Concurrency in Blob storage"](https://learn.microsoft.com/en-us/azure/storage/blobs/concurrency-manage). Mostly an exercise on understanding approaches. Knowing what we know now (as far as my understanding) - that whether or not an API call succeeds or not doesn't necessarily determine the outcome of what has actually happened on the storage - I'm under the impression that there must be a more proper way of handling all of this.

**Comment by danmason** [comment]
Following on from conversations yesterday on this process, we decided on the following:
- The risk of taking an interrupted exception in the `abort` process as "we can delete x file" - without knowing whether or not that file was already written/comitted by another node - doesn't seem particularly worth it.
- Given that this whole issue is derived from the writing of empty files to clear the staged blocks, we've decided on removing the whole "upload an empty file & delete" flow from aborting azure multiparts - any uncommitted blocks get automatically cleared when a new set of blocks get commited or after 7 days no activity regardless, so the impact is relatively small.
- Shall action this & do an azure run against it (the load phase still currently hangs - seems related to Kafka, so will file a separate issue).
